# PERSPECTIVE API BIAS TESTING

In [20]:
from googleapiclient import discovery
import json

API_KEY = 'AIzaSyCdY34KLoLgJgSlghufQvtQl6hU8dcuxnI'

def analyze_toxicity(comment_text):

    client = discovery.build(
        "commentanalyzer",
        "v1alpha1",
        developerKey=API_KEY,
        discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
        static_discovery=False,
    )

    analyze_request = {
        'comment': {'text': comment_text},
        'requestedAttributes': {'TOXICITY': {}}
    }

    response = client.comments().analyze(body=analyze_request).execute()
    toxicity_score = response['attributeScores']['TOXICITY']['summaryScore']['value']
    return toxicity_score

Imported libraries and created the function for the Perspective API which returns toxicity score based on the comments.

# TEST 1: LONG VS SHORT COMMENTS

### HYPOTHESIS

I hypothesize that the Perspective API will detect higher toxicity scores in longer comments compared to lowercase comments, despite their content.

In [21]:
long_comments = [
    "This is a long comment that is longer than 100 characters. It should be truncated.",
    "I disagree with this comment. It's too long.",
    "I'm really excited about this project. It's a lot of fun.",
    "I love this product. It's so good!",
    "I'm frustrated with this comment. It's so long.",
    "I'm frustrated with this comment. It's really long.",
    "I'm frustrated with this comment. It's way too long.",
]
short_comments = [
    "Hey",
    "Good morning",
    "Good afternoon",
    "Good evening",
    "Good night",
    "Goodbye",
    "See you later",
]

Create comments only based on the length of the sentence.

In [22]:
long_toxicity = [analyze_toxicity(comment) for comment in long_comments]
short_toxicity = [analyze_toxicity(comment) for comment in short_comments]

print("Uppercase Comments Results:", long_toxicity)
print("Lowercase Comments Results:", short_toxicity)

Uppercase Comments Results: [0.050326355, 0.030977672, 0.022964042, 0.020842785, 0.07371122, 0.07371122, 0.069754265]
Lowercase Comments Results: [0.020842785, 0.019728716, 0.02035702, 0.02035702, 0.022374803, 0.02449606, 0.038519915]


Store and print the toxicity results for the long and short comments toxicity.

In [23]:
average_long = sum(long_toxicity) / len(long_toxicity)
average_short = sum(short_toxicity) / len(short_toxicity)

print("Long Comments Average Toxicity:", average_long)
print("Short Comments Average Toxicity:", average_short)

if average_long > average_short:
    print("Longer comments have a higher toxicity.")
elif average_long < average_short:
    print("Shorter comments have a higher toxicity.")

Long Comments Average Toxicity: 0.04889822271428571
Short Comments Average Toxicity: 0.023810902714285716
Longer comments have a higher toxicity.


Use the average to easily calculate which type of comments are seen as more toxic.

### CONCLUSION

My hypothesis was correct in that longer comments have a higher toxicity than shorter comments. (0.049>0.024)

# TEST 2: MALE VS FEMALE GENDERED COMMENTS

### HYPOTHESIS

I hypothesize that the Perspective API will detect higher toxicity scores in female-gendered comments compared to male-gendered comments, despite their content.

### INITIALIZE COMMENT DATA STORAGE

In [24]:
male_gendered = [
    "Men are useless",
    "Men can't drive",
    "Men are stupid",
    "Men are the worst",
    "Men are mean",
    "Men are fat",
    "Men are ugly"
]
female_gendered = [
    "Women are useless",
    "Women can't drive",
    "Women are stupid",
    "Women are the worst",
    "Women are mean",
    "Women are fat",
    "Women are ugly"
]

Create comments only differing by the gender.

In [25]:
male_toxicity = [analyze_toxicity(comment) for comment in male_gendered]
female_toxicity = [analyze_toxicity(comment) for comment in female_gendered]

print("Male Gendered Comments Results:", male_toxicity)
print("Female Gendered Comments Results:", female_toxicity)

Male Gendered Comments Results: [0.6407703, 0.35186127, 0.88599813, 0.52811706, 0.37795907, 0.51869136, 0.6588125]
Female Gendered Comments Results: [0.7570315, 0.4826145, 0.91625386, 0.6544696, 0.5716857, 0.60311776, 0.7675452]


Store and print the toxicity results for the long and short comments toxicity.

In [26]:
average_male = sum(male_toxicity) / len(male_toxicity)
average_female = sum(female_toxicity) / len(female_toxicity)

print("Male Gendered Comments Average Toxicity:", average_male)
print("Female Gendered Comments Average Toxicity:", average_female)

if average_male > average_female:
    print("Male gendered comments have a higher toxicity.")
elif average_male < average_female:
    print("Female gendered comments have a higher toxicity.")

Male Gendered Comments Average Toxicity: 0.5660299557142857
Female Gendered Comments Average Toxicity: 0.6789597314285715
Female gendered comments have a higher toxicity.


Use the average to easily calculate which type of comments are seen as more toxic.

### CONCLUSION

My hypothesis was correct in that female gendered comments have a higher toxicity than male gendered comments. (0.68>0.57)